osbs/http.py

```diff
@@ -160,203 +79,68 @@ def __init__(self, url, method, data=None, kerberos_auth=False,
-        sel = self.curl_multi.select(SELECT_TIMEOUT)
-        if sel == -1:
-            raise OsbsException("CurlMulti.select() timed out")
-        elif sel == 0:
-            # sel==0 means curl_multi_fdset returned -1
-            # manual page suggests sleeping >100ms when this happens:(
-            time.sleep(0.1)
-
-    def _any_data_received(self):
-        return self.response_buffer.tell() != 0
+            args['stream'] = True
+
+        args['headers'] = headers
+        self.req = requests.request(method, url, **args)
+
+        self.headers = self.req.headers
+        self.status_code = self.req.status_code
 
     def _get_received_data(self):
-        result = self.response_buffer.getvalue()
-        self.response_buffer.truncate(0)
-        self.response_buffer.seek(0)
-        return self.response_decoder.decode(result, final=self.finished)
+        return self.req.text
 
     def iter_chunks(self):
-        while True:
-            self._perform()
-            if self._any_data_received():
-                yield self._get_received_data()
-            if self.finished:
-                break
-            self._select()
-
-        logger.debug("end of the stream")
-        self.close()
+        return self.req.iter_content(None)
 
     def iter_lines(self):
-        chunks = self.iter_chunks()
-        return self._split_lines_from_chunks(chunks)
-
-    @staticmethod
-    def _split_lines_from_chunks(chunks):
-        # same behaviour as requests' Response.iter_lines(...)
-
-        pending = None
-        for chunk in chunks:
-
-            if pending is not None:
-                chunk = pending + chunk
-            lines = chunk.splitlines()
-
-            if lines and lines[-1] and chunk and lines[-1][-1] == chunk[-1]:
-                pending = lines.pop()
-            else:
-                pending = None
-
-            for line in lines:
-                yield line
-
-        if pending is not None:
-            yield pending
-
-    @staticmethod
-    def _curl_debug(debug_type, debug_msg):
-        try:
-            logger_name = {
-                pycurl.INFOTYPE_TEXT: 'curl',
-                pycurl.INFOTYPE_HEADER_IN: 'in',
-                pycurl.INFOTYPE_HEADER_OUT: 'out'
-            }[debug_type]
-        except KeyError:
-            return
-
-        curl_logger = logging.getLogger(__name__ + '.' + logger_name)
-        for line in debug_msg.splitlines():
-            if not line:
-                continue
-            curl_logger.debug(line)
-
-    @property
-    def encoding(self):
-        encoding = None
-        if 'content-type' in self.headers:
-            content_type = self.headers['content-type'].lower()
-            match = re.search(r'charset=(\S+)', content_type)
-            if match:
-                encoding = match.group(1)
-        if encoding is None:
-            encoding = 'utf-8'  # assume utf-8
-
-        return encoding
+        return self.req.iter_lines(decode_unicode=True)
```

Comment:
This returns `bytes` not `str`, because the headers returned in the response include:
```
Transfer-Encoding: chunked, chunked
Content-Type: application/json
```

i.e. the charset is not specified. This leads to an exception when using "build-logs -f".

RFC 4627 defines a method for determining the encoding of an application/json response by examining the first four bytes, and this is implemented in `requests.utils.guess_json_utf()`.

It seems that it is not exactly http.py's job to do this decoding as part of the request object, but I think it would be useful for it to provide a function to do it, like this:
```
def decoded_json(iterable):
    """Return str objects for each JSON string in iterable

    See https://tools.ietf.org/html/rfc4627#section-3
    """
    for line in iterable:
        yield line.decode(requests.utils.guess_json_utf(line)
```
(needs a test case too)

and then core.py should use this function whenever it uses `iter_lines()` expecting JSON strings, like this:
```
for line in decoded_json(response.iter_lines()):
```